src/marker.rs

```diff
@@ -25,153 +364,302 @@ pub fn marker_2_rectancle(
-                }
-                let res = &local_listener.clone().lookup_transform(
-                    &msg.header.frame_id,
-                    &static_frame.clone(),
-                    msg.header.stamp,
-                );
-                match &res {
-                    Ok(res) => res,
-                    Err(_e) => return,
-                };
-
-                _lines.extend(marker_2_rectancle(
-                    &msg,
-                    &res.as_ref().unwrap().transform,
-                ));
-            },
-        )
-        .unwrap();
+    fn add_marker(&mut self, marker: &rosrust_msg::visualization_msgs::Marker) {
+        let transform = &self._tf_listener.clone().lookup_transform(
+            &marker.header.frame_id,
+            &self._static_frame.clone(),
+            marker.header.stamp,
+        );
+        match &transform {
+            Ok(transform) => transform,
+            Err(_e) => return,
+        };
 
-        MarkerListener {
-            config,
-            lines: blines,
-            _tf_listener: tf_listener,
-            _subscriber: _sub,
+        self._markers
+            .entry(marker.ns.clone())
+            .and_modify(|namespace| {
+                let res = parse_marker_msg(&marker, &transform.as_ref().unwrap().transform);
+                namespace.insert(res.id, res);
+            })
+            .or_insert_with(|| {
+                let res = parse_marker_msg(&marker, &transform.as_ref().unwrap().transform);
+                let mut namespace = BTreeMap::<i32, TermvizMarker>::new();
+                namespace.insert(res.id, res);
+                namespace
+            });
+    }
+
+    fn delete_marker(&mut self, marker_ns: String, marker_id: i32) {
+        self._markers.entry(marker_ns).and_modify(|namespace| {
+            namespace.remove(&marker_id);
+        });
+    }
+
+    fn clear(&mut self) {
+        self._markers.clear();
+    }
+
+    fn clear_namespace(&mut self, marker_ns: String) -> Vec<i32> {
+        let mut res = Vec::new();
+        self._markers.entry(marker_ns).and_modify(|namespace| {
+            for marker in namespace.values() {
+                res.push(marker.id);
+            }
+            namespace.clear();
+        });
+        res
+    }
+
+    fn get_lines(&self) -> Vec<Line> {
+        let mut res = Vec::<Line>::new();
+        for namespace in self._markers.values() {
+            for marker in namespace.values() {
+                res.extend(marker.lines.to_vec());
+            }
         }
+        res
     }
+}
 
-    pub fn get_lines(&self) -> Vec<Line> {
-        self.lines.read().unwrap().to_vec()
+///Class that handles the lifecycle of the markers.
+///
+/// Markers that provide a lifecycle (i.e. not 0) need to be deleted from the container
+/// once the lifecycle is reached. To do this we add another container which has a timer.
+/// The timer collects jobs after a delay to delete markers from the markers container.
```

Comment:
The guards collect and do stuff correct? The timer just runs

Reply:
If I understood the doc, the timer has the scheduled tasks. It has 2 threads, one to execute a task and the other to watch when the next one needs to be executed. The guard is just an object that cancels the task, it does not hold any task, the timer does.